source/common/quic/envoy_quic_client_session.cc

```diff
@@ -150,6 +150,17 @@ void EnvoyQuicClientSession::OnTlsHandshakeComplete() {
   streamInfo().upstreamInfo()->upstreamTiming().onUpstreamHandshakeComplete(
       dispatcher_.timeSource());
 
+  const auto quicConn = static_cast<EnvoyQuicClientConnection*>(quicConnection());
+  if (quicConn != nullptr) {
+    quicConn->connectionSocket()->connectionInfoProvider().maybeSetInterfaceName(
+        quicConn->connectionSocket()->ioHandle());
```

Comment:
I don't think setting this in the session layer is the right place. It should be done by the connection layer if possible.

Further thinking, since we set the interface name both for TCP and QUIC now(TCP done by https://github.com/envoyproxy/envoy/pull/19758). So it would be great to set the interface name in one place.

This also can avoid some bugs due to copy the connection around. (like https://github.com/envoyproxy/envoy/issues/17168 is trying to fix )

I guess we only can get the interface name after bind the socket (correct me if I'm wrong), so one place what I'm thinking is in the `SocketImpl::bind method` 
https://github.com/envoyproxy/envoy/blob/e85654e11c8e88e2b70ac15afa93573b26f19541/source/common/network/socket_impl.cc#L52

cc @mattklein123 @alyssawilk

Reply:
One can get network interfaces from kernel at any time but to pick one that matches the connection's local address which is retrieved via getsockname() we need to call maybeSetInterfaceName() on a bound socket.

However, a client socket doesn't necessarily always bind() before write(). For TCP, if the local address is not specified, kernel will pick the right network interface and a port for the connection. For QUIC, right now QUIC client connection always calls bind() because we will pick a network interface according to the remote address type if the local address is not provided via getLocalAddress(), see createConnectionSocket() in source/common/quic/envoy_quic_utils.cc. So it is infeasible to unify the place for TCP and QUIC to call maybeSetInterfaceName(). For QUIC, the most feasible place is after bind() in createConnectionSocket() if you want a quick fix.

Admittedly QUIC client socket creation should align with TCP, but even in that case, QUIC doesn't need to wait till handshake finish to call maybeSetInterfaceName() but after the connection receives the first packet. To do so, you will need to rewrite createConnectionSocket() and override some QuicConnection packet processing callbacks. Feel free to chose either approach.

BTW, thanks for contributing to Envoy H3!

Reply:
> 
thanks for your comments, will look into it, your guide is very improtant for me as a beginer for H3 and envoy, apperciated.

Reply:
> However, a client socket doesn't necessarily always bind() before write(). For TCP, if the local address is not specified, kernel will pick the right network interface and a port for the connection. 

Thanks! good to learn this.